src/libraries/System.Text.Json/gen/Reflection/TypeExtensions.cs

```diff
@@ -16,43 +14,60 @@ public static string GetCompilableName(this Type type)
                 return GetCompilableName(type.GetElementType()) + "[]";
             }
 
-            string compilableName;
-
-            if (!type.IsGenericType)
+            if (type.IsGenericParameter)
             {
-                compilableName = type.FullName;
+                return type.Name;
             }
-            else
-            {
-                StringBuilder sb = new();
 
-                string fullName = type.FullName;
-                int backTickIndex = fullName.IndexOf('`');
+            StringBuilder sb = new();
 
-                string baseName = fullName.Substring(0, backTickIndex);
+            sb.Append("global::");
 
-                sb.Append(baseName);
+            string @namespace = type.Namespace;
+            if (!string.IsNullOrEmpty(@namespace) && @namespace != JsonConstants.GlobalNamespaceValue)
+            {
+                sb.Append(@namespace);
+                sb.Append('.');
+            }
 
-                sb.Append('<');
+            int argumentIndex = 0;
+            AppendTypeChain(sb, type, type.GetGenericArguments(), ref argumentIndex);
 
-                Type[] genericArgs = type.GetGenericArguments();
-                int genericArgCount = genericArgs.Length;
-                List<string> genericArgNames = new(genericArgCount);
+            return sb.ToString();
 
-                for (int i = 0; i < genericArgCount; i++)
+            static void AppendTypeChain(StringBuilder sb, Type type, Type[] genericArguments, ref int argumentIndex)
+            {
+                Type declaringType = type.DeclaringType;
+                if (declaringType != null)
                 {
-                    genericArgNames.Add(GetCompilableName(genericArgs[i]));
+                    AppendTypeChain(sb, declaringType, genericArguments, ref argumentIndex);
+                    sb.Append('.');
                 }
+                int backTickIndex = type.Name.IndexOf('`');
+                if (backTickIndex == -1)
+                {
+                    sb.Append(type.Name);
+                }
+                else
+                {
+                    sb.Append(type.Name, 0, backTickIndex);
 
-                sb.Append(string.Join(", ", genericArgNames));
+                    sb.Append('<');
 
-                sb.Append('>');
+                    int startIndex = argumentIndex;
+                    argumentIndex = type.GetGenericArguments().Length;
```

Comment:
I'm not sure I follow this logic here. `argumentIndex` is related to `genericArguments` array but here you're using different array (could you be using `genericArguments` here?). Why does this need to be `ref` if you're always moving it to the end?

Reply:
should we be doing `+=` rather than `=` in that case? I assume that in the top most iteration you'd want to get first element, then second then third?